control/db.py

```diff
@@ -301,17 +347,51 @@ def delete_config(self):
     def restore(self, callbacks):
         """Restores gateway config to persistent config specifications."""
 
+        self.gateway_rpcs = callbacks
         omap_version = self._read_key(self.OMAP_VERSION_KEY)
         if omap_version == "1":
             self.logger.info("This omap was just created. Nothing to restore")
         else:
-            omap_dict = self._read_all()
-            self._restore_bdevs(omap_dict, callbacks[self.BDEV_PREFIX])
-            self._restore_subsystems(omap_dict,
-                                     callbacks[self.SUBSYSTEM_PREFIX])
-            self._restore_namespaces(omap_dict,
-                                     callbacks[self.NAMESPACE_PREFIX])
-            self._restore_hosts(omap_dict, callbacks[self.HOST_PREFIX])
-            self._restore_listeners(omap_dict, callbacks[self.LISTENER_PREFIX])
-            self.version = int(omap_dict[self.OMAP_VERSION_KEY])
+            omap = self._read_all()
+            self._update_caller(omap, True)
+            self.config = omap
+            self.version = int(omap[self.OMAP_VERSION_KEY])
             self.logger.info("Restore complete.")
+
+    def update(self,
+               notify_id=None,
+               notifier_id=None,
+               watch_id=None,
+               data=None):
+        """Checks for updated persistent config and initiates local update."""
+
+        omap_version = self._read_key(self.OMAP_VERSION_KEY)
+        if str(self.version) != omap_version:
+            omap = self._read_all()
+            # Handle OMAP additions
+            added_keys = omap.keys() - self.config.keys()
+            added = {key: omap[key] for key in added_keys}
+            self._update_caller(added, True)
+            # Handle OMAP removals
+            removed_keys = self.config.keys() - omap.keys()
+            removed = {key: self.config[key] for key in removed_keys}
+            self._update_caller(removed, False)
```

Comment:
I see potential issues with the order. For example, if a object is going to be removed which is currently still referenced by one of the other objects that only changed, it will fail to remove.